[Feat] 소셜 로그인 정리

her0807 opened the pull request:
알바가기 전까지 임시 작업한 것

her0807 committed:
feat: 간단한 social test
1)소셜 로그인 특성상 특정 API에 가서 값을 받아오고, 그 결과값(TOKEN)
을 받아서 인증 절차를 거쳐야 함. 그것은 이미 각 소셜 서비스를 구현하면서
확인 했기 때문에 테스트 코드를 작성하지않고, 로그인을 요청하는 URL을 만드는
메소드만 테스트 함.

her0807 committed:
feat: HttpMessageNotWritableException: No converter for 오류가 났지만 임시저장
거의 구현 다함..

hyeonic commented on src/main/java/com/skhuedin/skhuedin/config/BearerAuthInterceptor.java:
```diff
@@ -23,14 +23,16 @@ public boolean preHandle(HttpServletRequest request,
         System.out.println(">>> interceptor.preHandle 호출");
         String token = authExtractor.extract(request, "Bearer");
-        if (StringUtils.isEmpty(token)) {
+        //값이 있으면 토큰 값을 저장한다.
+
+        if (StringUtils.isEmpty(token)) { // 값이 비어있으면 ture 를 반환한다.
             return true;
```
StringUtils보다 Java 11을 사용하고 있기 때문에 String 클래스에 선언된 isEmpty를 사용해보는 건 어떨까요? 제안드립니다.

[stringutils-isblank-vs-string-isempty](https://stackoverflow.com/questions/23419087/stringutils-isblank-vs-string-isempty)

hyeonic commented on src/main/java/com/skhuedin/skhuedin/controller/UserApiController.java:
```diff
@@ -0,0 +1,42 @@
+import com.skhuedin.skhuedin.dto.user.UserMainResponseDto;
+import com.skhuedin.skhuedin.service.UserService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.http.*;
```
직관적인 import를 확인하기 위해 * 수정 부탁드려요!

hyeonic reviewed (commented):
추가적인 리뷰는 완성되면 다시 한번 살펴보겠습니당

her0807 committed:
feat: 로그인 로직 수정
1) 원래 preHandle 에서 토큰 값이 있으면 바로 로그인 시켜주려고 했는데
토큰을 발급 받은 이후에 데이터가 업데이트 될 가능성도 있다고 판단해서
토큰이 있으면 prehandle 을 토큰을 발급하지 않고 기존에 가지고 있는 토큰이 유효한지 검증하고
유효하면 그값을 넘긴다.

없으면 토큰을 발급해서 넘긴다.

2) 소셜로그인을 요청하면 결과값을 바로 우리 유저 token과 상태를 header 에 넣어주고
 CommonResponse 에 User 정보를 담아서 반환하도록 설계했다.

hyeonic commented on src/main/java/com/skhuedin/skhuedin/config/BearerAuthInterceptor.java:
```diff
@@ -21,11 +24,11 @@ public BearerAuthInterceptor(AuthorizationExtractor authExtractor, JwtTokenProvi
         //값이 있으면 토큰 값을 저장한다.
 
-        if (StringUtils.isEmpty(token)) { // 값이 비어있으면 ture 를 반환한다.
+        if (Strings.isEmpty(token)) { // 값이 비어있으면 ture 를 반환한다.
```
제가 말씀드린 것은 String 객체인 token을 활용하는 것 이었어요!
```
token.isEmpty();
```

hyeonic commented on src/main/java/com/skhuedin/skhuedin/service/UserService.java:
```diff
@@ -57,25 +57,37 @@ public String getToken(String Token) {
     }
 
-    public String signIn(User user) {
+    public User login(UserMainResponseDto user) { // 회원가입
+        User findUser = userRepository.findByEmail(user.getEmail())
+                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 유저"));
+        return findUser;
+    }
```
login의 목적은 무엇인가요? 주석은 회원가입이라고 되어있는데 값을 저장하는 로직은 없어보입니다! 단순히 userDto를 활용하여 저장하는 것은 save의 역할이 될 수 있지 않을까요? 혹은 단순히 조회의 목적인가요? 메소드의 역할이 다소 헷갈리는 부분이 있습니다.

hyeonic commented on src/main/java/com/skhuedin/skhuedin/service/UserService.java:
```diff
@@ -57,25 +57,37 @@ public String getToken(String Token) {
         if (!findUser.getPassword().equals(user.getPassword()))
             throw new IllegalArgumentException("암호 불일치");
+
+        // 로그인 전 변경 사항이 있는지 체크
+        if (!checkUpdate(findUser, user)) {
+            update(findUser.getId(), user);
+            return createToken(findUser.getEmail());
+        }
```
단순히 if 를 활용하여 값을 비교하는 것으로 변경 사항을 체크해야 할 필요성이 있는지 생각해보면 좋을 것 같아요! update 메소드를 실행하게 되면 transaction 안에서 변경이 감지되면 (즉, 변경된 값이 들어오게 되면) 자동으로 update 쿼리를 실행하고, 값이 동일하면 실행하지 않는 걸로 알고 있습니다. 이 부분은 테스트 코드를 활용하여 값이 변경됬을 때와 되지 않았을 때를 비교하여 차이를 비교해보는 것이 어떨까요?

hyeonic commented on src/main/java/com/skhuedin/skhuedin/social/kakao/KakaoOauth.java:
```diff
@@ -108,9 +110,15 @@ public String requestAccessToken(String code) {
+        //유저 형식에 맞게 저장하기
+        user = saveKakaoUser(kakaoProfile);
+
+        //회원 가입으로 보내기
+        userService.signUp(user);
```
회원의 유무를 파악하는 것은 userService의 일이 되는 건가요? 쭉 살펴보면 무조건적으로 회원가입으로 보내야 한다는 생각이 듭니다. 단순히 user를 return 하면 userService는 return한 user를 기반으로 로직을 처리할 수 있지 않을까란 생각이 들어요!

hyeonic reviewed (commented):
전반적으로 다시 한번 살펴보았습니다! 전체적인 큰 그림이 머리속에는 그려지는데 약간 의아한 부분이 몇 군데 있어서 comment 남겼습니다! 추가적으로 각 파일마다 애노테이션 부분까지 마지막으로 확인 부탁드려요! 아직까지도 *로 표기된 부분이 많은 것 같습니다.

her0807 committed:
feat: userDTO활용, 불필요한 로직 수정
1)user dto 를 만들어 활용, 엔티티를 외부로 노출하지 않아 변경을 최소화 하기 위함
2)불필요한 service 제거
3)import.* 변경 후 전체 명시

hyeonic commented on src/main/java/com/skhuedin/skhuedin/dto/follow/FollowMainResponseDto.java:
```diff
@@ -0,0 +1,25 @@
+    public Follow toEntity(Follow follow) {
+        return Follow.builder()
+                .user(this.user)
+                .build();
```
Response를 위한 Dto인데 엔티티로 바꿔주는 이유가 있을까요? Dto의 역할이 모호한 것 같아요!
request의 경우 Dto -> entity 변환의 역할을 담당하고, Response의 경우 한정된 정보를 제공하기 위해 
entity -> Dto로 변환하는 역할을 담당해야 하는게 아닐지 남겨봅니다!

hyeonic commented on src/main/java/com/skhuedin/skhuedin/dto/user/UserMainResponseDto.java:
```diff
@@ -1,18 +1,44 @@
+                .graduationYear(this.graduationYear)
+                .build();
+        return user;
     }
```
위와 동일합니다. 지금 ResponseDto가 두 개의 역할을 모두 다 하고 있어요! 좀 더 고민의 필요성이 있어보입니다.

hyeonic commented on src/main/java/com/skhuedin/skhuedin/social/OauthController.java:
```diff
@@ -31,17 +43,30 @@ public void socialLoginType(
+        header.add("Authorization", "Bearer " + userService.signIn(user));
 
-        return ResponseEntity.ok().body(new TokenResponse(token, "bearer"));
+        return new ResponseEntity<>(new CommonResponse<>(user), header, HttpStatus.OK);
```
후에 프론트와 통신을 위해서는 header에 값을 채워주기 보단 response body에 token 값을 채워주고 그것을 활용하는 방안으로 생각하면 어떨까요? Authorization을 설정해 주는 부분은 프론트의 역할이라고 생각합니다.

hyeonic commented on src/main/java/com/skhuedin/skhuedin/social/OauthService.java:
```diff
@@ -29,16 +31,23 @@ public void request(Provider socialLoginType) {
         SocialOauth socialOauth = this.findSocialOauthByType(socialLoginType);
         String redirectURL = socialOauth.getOauthRedirectURL();
         try {
+            //페이지를 아래 주소로 리다이렉트 지시
             response.sendRedirect(redirectURL);
```
service layer에서 controller의 역할을 하고 있는 것 같아 보이는데 어떻게 생각하시나요? controller와 service의 역할이 모호한 것 같아요.

hyeonic reviewed (commented):
Dto의 사용에 의문에 생겨 추가 comment 남깁니다. 추가적으로 각 social에 save 메소드의 경우 중복되는 부분이 많은데 userService 혹은 socialService에서 공통으로 로직을 통일하면 어떨까 제안 드립니다!

her0807 committed:
feat: respones dto, request dto 변경
1) dto 의 명확한 활용을 위한 분리 시행
2) token 을  response에 담기 위한 추가 클래스 생성
3) service 와 controller 의 명확한 분리 피드백 반영

hyeonic commented on src/main/java/com/skhuedin/skhuedin/dto/follow/FollowSaveRequestDto.java:
```diff
@@ -0,0 +1,25 @@
+        this.user = user;
+    }
+
+    public Follow toEntity(Follow follow) {
+        return Follow.builder()
+                .user(this.user)
+                .build();
```
```
public Follow toEntity() {
    return Follow.builder()
        .user(this.user)
        .build();
```
인자로 들어오는 Follow는 전혀 사용되고 있지 않아 생략해도 괜찮을 것 같습니다.

추가적으로 follow save를 위한 requestDto인데 fromUser와 toUser로 나눠서 필드를 추가하기로 하지 않았었나요?

hyeonic commented on src/main/java/com/skhuedin/skhuedin/service/UserService.java:
```diff
@@ -22,10 +23,8 @@ public Long save(User user) {
-        User findUser = userRepository.findById(id).orElseThrow(() ->
-                new IllegalArgumentException("해당 user 가 존재하지 않습니다. id=" + id));
-        findUser.update(userMainResponseDto.toEntity());
+    public void update(User user, UserSaveRequestDto requestDto) {
+        user.update(requestDto.toEntity());
```
만약 User user 부분에서 영속성 컨텍스트에 포함되지 않은 user가 들어온다면 어떻게 대응해야 하나요? user의 경우 지금 값을 업데이트하여도 제대로 된 반영이 되지 않을 것으로 예상되는데 확인 해보셨나요??

hyeonic reviewed (commented):
우선 보이는 부분 comment 남겼습니다! 특히 follow 부분 회의 나눈 결과가 아직 반영이 안된 것 같아요

her0807 committed:
feat: token 로직 수정 + role 추가
1) user role 추가해서 권한 확인 로직을 preHandle 에 추가
2) 불필요하게 쿼리문이 여러번 조회 됨을 발견, 정정

hyeonic commented on src/main/java/com/skhuedin/skhuedin/config/BearerAuthInterceptor.java:
```diff
@@ -12,32 +18,69 @@
-
-    public BearerAuthInterceptor(AuthorizationExtractor authExtractor, JwtTokenProvider jwtTokenProvider) {
-        this.authExtractor = authExtractor;
-        this.jwtTokenProvider = jwtTokenProvider;
-    }
+    private UserService userService;
```
private final이 생략되어도 괜찮은 건가요??

hyeonic commented on src/main/java/com/skhuedin/skhuedin/config/BearerAuthInterceptor.java:
```diff
@@ -12,32 +18,69 @@
+        HandlerMethod handlerMethod = (HandlerMethod) handler;
 
+        // @MyRole 받아오기
+        MyRole mySecured = handlerMethod.getMethodAnnotation(MyRole.class);
```
클래스이름과 통일성을 주면 좋을 것 같아요

hyeonic commented on src/main/java/com/skhuedin/skhuedin/config/BearerAuthInterceptor.java:
```diff
@@ -12,32 +18,69 @@
+        //값이 있으면 토큰 값을 저장한다.
+        String token = authExtractor.extract(request, "Bearer");
+
+        // 값이 비어있으면 ture 를 반환한다.
+        if (token.isEmpty()) {
+            return true;
+        }
```
@MyRole로 인하여 권한에 대한 인증이 필요하다면 token 값이 반드시 존재해야 하지 않을까요?

hyeonic commented on src/main/java/com/skhuedin/skhuedin/controller/UserApiController.java:
```diff
@@ -1,16 +1,40 @@
+        return ResponseEntity.status(HttpStatus.OK).body(new CommonResponse<>(responseDto));
+    }
+
+    @GetMapping("users{id}")
+    public ResponseEntity<? extends BasicResponse> findAll() {
+        List<UserMainResponseDto> users = userService.findAll();
+        return ResponseEntity.status(HttpStatus.OK).body(new CommonResponse<>(users));
+    }
```
@GetMapping 부분 오타가 있는 것 같습니다

hyeonic commented on src/main/java/com/skhuedin/skhuedin/service/UserService.java:
```diff
@@ -59,19 +65,35 @@ public String signUp(UserSaveRequestDto requestDto) { // 회원가입
     }
 
     public User findByEmail(String email) {
+        Optional<User> user = userRepository.findByEmail(email);
+        // 값이 없는데 get 하면 오류가 나기 때문에 isPresent 로 잡아줌,
+        if (!user.isPresent())
+            return null; // Bearer 검증에 null 값을 넘기기 위해 일부러 이렇게 작성함
+        else return user.get();
+    }
```
Optional을 사용할 때 isPresent()-get()보다 orElse(), orElseGet(), orElseThrow()를 사용해보는 것은 어떨까요?

[Java Optional 바르게 쓰기](http://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/)

hyeonic commented on src/main/java/com/skhuedin/skhuedin/service/UserService.java:
```diff
@@ -59,19 +65,35 @@ public String signUp(UserSaveRequestDto requestDto) { // 회원가입
+        if (!user.isPresent())
             return null;
-        else return userRepository.findByEmail(email).get();
+        else return user.get();
```
위와 동일합니다.

hyeonic commented on src/main/java/com/skhuedin/skhuedin/social/kakao/KakaoOauth.java:
```diff
@@ -109,6 +110,8 @@ public UserSaveRequestDto requestAccessToken(String code) {
         //유저 형식에 맞게 저장하기
         user = saveKakaoUser(kakaoProfile);
 
+        log.info(response2.getBody());
```
디버깅하고 난 후 필요가 없다면 지우셔도 될 것 같아요~

hyeonic reviewed (commented):
코드 리뷰 남겼습니다ㅏ

her0807 committed:
feat: 피드백 반영
1) Optional orElse 수정
2) User Controller 오타 수정
